### PR TITLE
Implement cache

### PR DESCRIPTION
**Pull request by marctemp**
https://dev.azure.com/defragovuk/DEFRA-FFC/_workitems/edit/127815

### CODE REVIEW

File: app/server.js [hunk 1]
```diff
@@ -1,13 +1,28 @@
-require('./insights').setup()
 const Hapi = require('@hapi/hapi')
+const config = require('./config')
+const catbox = config.useRedis ? require('@hapi/catbox-redis') : require('@hapi/catbox-memory')
+const catboxOptions = config.useRedis ? config.catboxOptions : {}
```
Comment (johnwatson484) [inline app/server.js]: We could determine the catbox config in the `config` module when we're setting it up to begin with?

File: app/server.js [hunk 2]
```diff
@@ -1,13 +1,28 @@
...
+
+require('./insights').setup()
```
Comment (johnwatson484) [inline app/server.js]: This should always be the first line executed on start up so we capture telemetry of any issues.  Should be first line in `index.js`.

File: app/get-sommit.js [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+      console.log('Using existing cache value')
+      return res
+    }
+    return 'mock read through cache method to be created and called'
+  } catch {
+    return undefined
+  }
+}
+
+module.exports = getSommit
```
Comment (johnwatson484) [inline app/get-sommit.js:16]: Does this need to be here still or should it be removed before merging?

File: app/routes/cache.js [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+  }
+},
+{
+  method: 'GET',
+  path: '/set',
+  handler: async (request, h) => {
+    await set(request, 'Web', `Value gotten from web get with timestamp ${new Date().toString()}`)
+    return h.response('ok').code(200)
+  }
+}]
```
Comment (johnwatson484) [inline app/routes/cache.js:19]: Do these routes need to exist?

File: app/cache/get.js [hunk 5]
```diff
@@ -0,0 +1,13 @@
+const getCache = require('./get-cache')
+const getCacheValue = require('./get-cache-value')
+
+const get = async (request, key) => {
+  try {
+    return await getCacheValue(getCache(request), key)
```
Comment (johnwatson484) [inline app/cache/get.js:6]: Can remove `await` here.  It's an async function so calling functions can await.

File: app/cache/get-cache-value.js [hunk 6]
```diff
@@ -0,0 +1,10 @@
+const getCacheValue = async (cache, key) => {
+  console.log(`Getting cache value for key: ${key}`)
+  try {
+    return await cache.get(key)
```
Comment (johnwatson484) [inline app/cache/get-cache-value.js:4]: Can remove await here.

File: app/cache/drop-cache-key.js [hunk 7]
```diff
@@ -0,0 +1,11 @@
+const dropCacheKey = async (cache, key) => {
+  console.log(`Dropping cache key: ${key}`)
+  try {
+    return await cache.drop(key)
```
Comment (johnwatson484) [inline app/cache/drop-cache-key.js]: No need to await

File: docker-compose.yaml [hunk 8]
```diff
@@ -2,9 +2,16 @@ services:
...
     container_name: ffc-pay-statement-receiver
+    depends_on:
+      - ffc-pay-statement-receiver-redis
     environment:
       APPINSIGHTS_CLOUDROLE: ffc-pay-statement-receiver-local
       APPINSIGHTS_INSTRUMENTATIONKEY: ${APPINSIGHTS_INSTRUMENTATIONKEY}
+      REDIS_HOSTNAME: ffc-pay-statement-receiver-redis
+      REDIS_PASSWORD: redis
+
+  ffc-pay-statement-receiver-redis:
```
Comment (johnwatson484) [inline docker-compose.yaml:14]: As this is used in testing, could it be moved to the `override` file.  Also suggest binding it's port to localhost so Redis clients can be used to interact with data.

Also suggest binding a Docker Volume so cache can be persisted between starts.

Something like this:

```
    volumes:
      - redis_data:/data
    ports:
      - "6379:6379"
```

Though pick a non-clashing port with other payment services that use Redis, which I think is only Request Editor.

**Review by johnwatson484 — COMMENTED** [review]
Couple of minor suggestions

File: app/cache/get.js (hunk 5 above)
Comment (marctemp) [reply app/cache/get.js:6]: Can't remove this -- I also thought same as yerself becuz its in an `async` func but a las, I cannot

File: app/cache/get-cache-value.js (hunk 6 above)
Comment (marctemp) [reply app/cache/get-cache-value.js:4]: Can't remove this -- I also thought same as yerself becuz its in an `async` func but a las, I cannot

File: app/get-sommit.js (hunk 3 above)
Comment (marctemp) [reply app/get-sommit.js:16]: Happy to remove these routes and mock func if you're happy with implementation?

File: app/cache/get.js (hunk 5 above)
Comment (johnwatson484) [reply app/cache/get.js:6]: Interesting, wonder why.  Could it be because there's a nested non-async function?  Doesn't feel that would be the reason though.

Comment (marctemp) [reply app/cache/get.js:6]: Yeah wouldn't explain not being able to do like-wise for `return await cache.get(key)` though

File: app/routes/cache.js (hunk 4 above)
Comment (marctemp) [reply app/routes/cache.js:19]: Handing over to Simon and note on this.

TL;DR I suggest keeping these routes in the main branch to help test in DEV and TEST.

This isn't currently in DEV or TEST due to missing ACR instances which I _assume_ will come when it merges into `main` <-- 1st thing to confirm.

Anyway, because of this I'm thinking these should be kept in to allow this (with the routes) to be elevated to DEV where we can use these to check the caching is working as it did in SND (same Redis config and instance so quietly confident). Just means this will give us an easy way to test in these environments.

Then in a follow-up ticket which is to basically merge Abi's (API returning blob statement) and mine (caching) work pieces together we can remove the routes then?

Comment (johnwatson484) [reply app/routes/cache.js:19]: Yes that makes sense.

File: docker-compose.override.yaml [hunk 9]
```diff
@@ -7,9 +7,21 @@ services:
...
     volumes:
       - ./app:/home/node/app
       - ./package.json:/home/node/package.json
+
+  ffc-pay-statement-receiver-redis:
+    ports:
+      - "6378:6379"
+    volumes:
+      - redis_data:/data
+    command: redis-server --dir /data --save 5 1
```
Comment (johnwatson484) [inline docker-compose.override.yaml:24]: What does this do?

File: docker-compose.test.yaml [hunk 10]
```diff
@@ -7,10 +7,22 @@ services:
...
     volumes:
       - ./app:/home/node/app
       - ./test:/home/node/test
       - ./test-output:/home/node/test-output
       - ./jest.config.js:/home/node/jest.config.js
       - ./package.json:/home/node/package.json
+
+  ffc-pay-statement-receiver-redis:
+    ports:
+        - "6378:6379"
```
Comment (johnwatson484) [inline docker-compose.test.yaml]: No need to expose ports for test container.  Will likely cause issues in Jenkins trying to bind to Jenkins localhost, particularly if port already in use.

File: docker-compose.test.yaml [hunk 11]
```diff
@@ -7,10 +7,22 @@ services:
...
+    volumes:
+      - redis_data:/data
+    command: redis-server --dir /data --save 5 1
+
+volumes:
+  redis_data:
```
Comment (johnwatson484) [inline docker-compose.test.yaml]: No need to bind persistent volume in tests

File: docker-compose.yaml [hunk 12]
```diff
@@ -2,9 +2,14 @@ services:
   ffc-pay-statement-receiver:
     build:
       context: .
-      target: production
+      target: development
```
Comment (johnwatson484) [inline docker-compose.yaml]: This should be production so build in pipeline targets production Dockerfile stage

Comment (marctemp) [reply docker-compose.yaml]: If this is production does in config `value.env` not become `production` which causes the `isProd` to be true which causes Redis to attempt to load with TLS which breaks it?

File: docker-compose.override.yaml (hunk 9 above)
Comment (marctemp) [reply docker-compose.override.yaml:24]: not sure if I need `--dir /data` which tells Redis where to spit its data backups (think its default, lemme find out)

`--save 5 1` is how often Redis makes snapshots of its content. 5 refers to the number of seconds and 1 is the number of item updates or creations.

This does it every time there is a change -- being local I thought it wouldn't matter that much.

Default was sommit like: once an day or 100+, once an hour or 1k+, once a minute or 10k+

File: docker-compose.yaml (hunk 12 above)
Comment (johnwatson484) [reply docker-compose.yaml]: `production` here relates to the Docker multi-stage build in `Dockerfile`.

Nothing to do with the Node environment.

image

Basically they're pretty much the same but use different parent images as a base and have a different default command.

The `docker-compose.override.yaml` file switches this to `development` locally anyway.

In Jenkins it builds without the override, so if you changed the default to `development` the image would be deployed with the incorrect parent image and not include the Palo Alto certificates or use the correct start up command.

File: docker-compose.override.yaml (hunk 9 above)
Comment (johnwatson484) [reply docker-compose.override.yaml:24]: I see, didn't realise you could configure `--save`.  That's cool :)

**Review by johnwatson484 — APPROVED** [review]
